### PR TITLE
[JSC] Validate funcref from JS function

### PR DESCRIPTION
**Pull request by Constellation**
#### 9fbb6a1b55afd7c1d3a4fff33578dd334242bb6c
<pre>
[JSC] Validate funcref from JS function
<a href="https://bugs.webkit.org/show_bug.cgi?id=250732">https://bugs.webkit.org/show_bug.cgi?id=250732</a>
rdar://104332115

Reviewed by Justin Michaud and Mark Lam.

This patch fixes funcref validation from JS function. We need to ensure that funcref value from JS
is actually Wasm functions. Fortunately, our use of funcref from wasm is always validating them, so,
it is just a semantic bug.

* JSTests/wasm/stress/funcref-validation.js: Added.
(async let):
* JSTests/wasm/stress/resources/funcref-validation.wasm: Added.
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
(JSC::Wasm::wasmToJS):

Canonical link: <a href="https://commits.webkit.org/259011@main">https://commits.webkit.org/259011@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/8309840e75de59f1b0f1dcf8dafdc16e3e4aa74c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103642 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12758 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36592 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112873 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/173204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13781 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3655 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/95885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/112029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109414 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/10621 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/93681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/95885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/92446 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25285 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/95885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/93782 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6128 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/26685 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/90209 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/3890 "Built successfully and passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/6305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/3212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29807 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/12287 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46198 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/98814 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8062 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/24875 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3292 "Built successfully and passed tests") | | | |

### CODE REVIEW

File: Source/JavaScriptCore/wasm/WasmOperations.cpp [hunk 1]
```diff
@@ -609,10 +627,17 @@ JSC_DEFINE_JIT_OPERATION(operationIterateResults, void, (Instance* instance, con
...
         default: {
-            if (isFuncref(returnType) || isExternref(returnType)) {
-                if (isFuncref(returnType) && !value.isCallable()) {
-                    throwTypeError(globalObject, scope, "Funcref value is not a function"_s);
-                    return;
+            if (isExternref(returnType))
+                ASSERT(returnType.isNullable());
+            else if (isFuncref(returnType)) {
+                ASSERT(returnType.isNullable());
+                if (isFuncref(returnType)) {
```
Comment (MenloDorian) [inline Source/JavaScriptCore/wasm/WasmOperations.cpp]: This is redundant: we already checked it at line 632.

File: Source/JavaScriptCore/wasm/WasmOperations.cpp [hunk 2]
```diff
@@ -544,6 +544,24 @@ JSC_DEFINE_JIT_OPERATION(operationConvertToF32, float, (Instance* instance, Enco
...
+    CallFrame* callFrame = DECLARE_WASM_CALL_FRAME(instance);
+    VM& vm = instance->vm();
+    JSGlobalObject* globalObject = instance->owner<JSWebAssemblyInstance>()->globalObject();
+    NativeCallFrameTracer tracer(vm, callFrame);
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    JSValue value = JSValue::decode(v);
+    WebAssemblyFunction* wasmFunction = nullptr;
+    WebAssemblyWrapperFunction* wasmWrapperFunction = nullptr;
+    if (!isWebAssemblyHostFunction(value, wasmFunction, wasmWrapperFunction) && !value.isNull()) {
```
Comment (MenloDorian) [inline Source/JavaScriptCore/wasm/WasmOperations.cpp]: UNLIKELY?

**Review by justinmichaud — APPROVED** [review]
r=me

**Review by MenloDorian — APPROVED** [review]
r=me with comments.

File: Source/JavaScriptCore/wasm/WasmOperations.cpp (hunk 1 above)
Comment (Constellation) [reply Source/JavaScriptCore/wasm/WasmOperations.cpp]: Fixed.

File: Source/JavaScriptCore/wasm/WasmOperations.cpp (hunk 2 above)
Comment (Constellation) [reply Source/JavaScriptCore/wasm/WasmOperations.cpp]: Fixed.

**Pull request merged**